Content translation: validate uploads

rafpaf opened the pull request:
This PR adds logic to validate content translation dictionary uploads and display errors to the user.

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/utGnP2ljFMMNOgi82vVn/c0813eb0-67a4-4b20-b746-0c3c780d08bb.png)

A valid dictionary must have:

* Three fields per row
* A valid locale in the first column
* One translation per string per locale (for example, you can only provide one translation of "Category" into German - you can't have two rows that begin "de,Category,")
* Cells containing strings no longer than 255 characters

rafpaf committed:
This introduces logic that validates content translation dictionaries
when they are uploaded.

A valid dictionary must have:

* Three fields per row
* A valid locale in the first column
* One translation per string per locale (for example, you can only provide one translation of "Category" into German - you can't have two rows that begin "de,Category,")
* Cells containing strings no longer than 255 characters

rafpaf committed:
Restore BE validation error logic

rafpaf commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/57544?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#57544** <a href="https://app.graphite.dev/github/pr/metabase/metabase/57544?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/metabase/metabase/57544?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#56880** <a href="https://app.graphite.dev/github/pr/metabase/metabase/56880?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#57481](https://github.com/metabase/metabase/pull/57481) <a href="https://app.graphite.dev/github/pr/metabase/metabase/57481?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#56879** <a href="https://app.graphite.dev/github/pr/metabase/metabase/56879?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#57534** <a href="https://app.graphite.dev/github/pr/metabase/metabase/57534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

rafpaf committed:
Don't verify summaries

rafpaf committed:
Fix missing import

sloansparger reviewed (approved):
FE looks good ignoring the lint errors.

sloansparger commented on enterprise/frontend/src/metabase-enterprise/content_translation/components/ContentTranslationConfiguration.tsx:
```diff
@@ -77,17 +100,19 @@ const UploadForm = () => {
         });
     },
-    [uploadContentTranslationDictionary, setStatus],
+    [uploadContentTranslationDictionary, setErrorMessages, setStatus],
```
I don't think setErrorMessages is a required dep

rafpaf committed:
Fix type issue

rafpaf committed:
Fix lint issues

dpsutton commented on enterprise/backend/src/metabase_enterprise/content_translation/api/dictionary.clj:
```diff
@@ -4,12 +4,49 @@
+(defn- collect-locale-error
+  "Returns an error message if a row does not have a valid locale."
+  [row-index locale]
+  (when (and (not (str/blank? locale))
```
no need to check the string for being blank

```clojure
pr-review=> (metabase.util.i18n/available-locale? "fr")
true
pr-review=> (metabase.util.i18n/available-locale? "")
false
pr-review=> (metabase.util.i18n/available-locale? nil)
false
```

edpaget commented on enterprise/backend/src/metabase_enterprise/content_translation/api/dictionary.clj:
```diff
@@ -4,12 +4,49 @@
+(defn- row-has-correct-number-of-fields
+  "Checks if a row has the expected format with exactly 3 columns."
+  [row]
+  (and (vector? row) (= (count row) 3)))
```
`count` will work with any sequential so I don't think you need to do the `vector?` test unless there's somewhere you're doing index based access.

johnswanson commented on enterprise/backend/src/metabase_enterprise/content_translation/api/dictionary.clj:
```diff
@@ -29,19 +66,61 @@
+(defn- validate-row
+  "Validate a single row of CSV data, recording errors and updating the set of seen translation keys."
+  [row-index row seen-keys]
+  (let [errors (transient [])
```
I'm curious if the use of `transient`s here and elsewhere is because of performance considerations that we've identified and profiled or is this pre-emptive?

Generally see transients as a good option to speed up hot spots but generally lead to less idiomatic clojure code, and here it seems like the number of changes we're making to the `errors` transient is bounded and relatively small, so I think I'd lean towards something more idiomatic.

`(remove nil? [(collect-duplication-error ...) (collect-locale-error ...) (collect-string-length-error ...) ...])` for example would just collect the non-nil error results (I guess you should verify that we're actually returning `nil` in the non-error case, instead of `false` or something)

rafpaf replied:
Thanks for the feedback! I've taken out the transients.

johnswanson commented on enterprise/backend/src/metabase_enterprise/content_translation/api/dictionary.clj:
```diff
@@ -29,19 +66,61 @@
+
+      ; Validate all rows before proceeding
+      (loop [rows (map-indexed vector csv-data)
+             errors (transient [])
```
Similar to above, maybe something like

```
    (let [csv-data (read-csv-file reader)
          errors (seq
                  (:errors (reduce (fn [{:keys [seen-keys] :as m} [row-index row]]
                                     (let [{row-errors :errors updated-seen-keys :seen-keys} (validate-row row-index row seen-keys)]
                                       (-> m
                                           (update :seen-keys #(apply conj %1 %2) updated-seen-keys)
                                           (update :errors #(apply conj %1 %2) row-errors))))
                                   {:seen-keys #{}
                                    :errors []}
                                   (map-indexed vector csv-data))))
          _ (when (seq errors)
```

would be more idiomatic, less side-effect-y, maybe worth thinking about.

rafpaf replied:
🙏 I've implemented your idea

rafpaf committed:
Address feedback:
* Don't do blank check
* Don't check msgid length or msgstr length
* Don't impose maximum length on msgid or msgstr
* Refactor, avoiding transients

rafpaf commented on enterprise/backend/src/metabase_enterprise/content_translation/api/dictionary.clj:
```diff
@@ -4,12 +4,49 @@
 
+(def ^:private http-status-unprocessable 422)
+
+(def ^:private max-string-length 255)
```
Good catch. I've taken out the max-string-length check.

rafpaf committed:
Move content translation model to OSS

rafpaf committed:
Fix config.edn

rafpaf committed:
Move some files around

rafpaf committed:
Fix some kondo errors

dpsutton committed:
Content translation cleanup (#58553)

* remove full response for simple map

```diff
-  {:status http-status-ok
-   :headers {"Content-Type" "application/json"}
-   :body (json/encode {:success true})})
+  {:success true})
```
this is equivalent and happens automatically

* just pass a file object

gonna change this as i don't love the api but it's better than the map
passing in the filename that wasn't needed. This is single argument

* pass rows in, not a file

* bit of cleanup in the dictionary creation namespace

johnswanson commented on src/metabase/content_translation/models.clj:
```diff
@@ -1,4 +1,4 @@
-(ns metabase-enterprise.content-translation.models
+(ns metabase.content-translation.models
```
Something I don't understand - why is it necessary that this be OSS?

rafpaf replied:
Thank you, that's a good catch. I'll move everything back into `enterprise`

johnswanson commented on enterprise/backend/src/metabase_enterprise/content_translation/dictionary.clj:
```diff
@@ -0,0 +1,91 @@
+                                                 :msgid  (str/trim msgid)
+                                                 :msgstr (str/trim msgstr)}
+                  errors                        (cond-> (row-errors state index translation)
+                                                  (seq extra) (conj (wrong-row-shape index)))]
```
more readable version of the blow imo:

```
(cond-> state
 true (update :seen ...)
 true (update :translations ...)
 (seq errors) (update :errors ...))
```

rafpaf commented on enterprise/backend/src/metabase_enterprise/content_translation/dictionary.clj:
```diff
@@ -0,0 +1,91 @@
+                      {:status-code 422
+                       :errors errors})))
+    ;; remove bad msgstrs after error generator for line number reporting reasons
+    (let [usable-rows (filter (comp is-msgstr-usable :msgstr) translations)]
```
If there are rows in the uploaded dictionary that don't have translations, like this, we just ignore them:

```
de,Bird,
```

Here's why this is important. Suppose the user downloads a template dictionary (which might have hundreds of strings, all without any translations), then adds a few translations, leaving the rest blank, and uploads this file. We should ignore the rows without translations.

rafpaf committed:
Content translation: impose maximum size on upload (#57666)
